### PR TITLE
Create StateT type

### LINKED ISSUE #14: Create StateT type
This implementation will be a simplification as we don't have HKTs yet

### PR DESCRIPTION
**Pull request by arturogutierrez**
Fixes #14 

Super basic example of State Monad:

```kotlin
// State for a random 6-dice roll
val dice = State<Random, Int> { r -> Pair(r, r.nextInt(6) + 1) }
// Applying a +3 modifier to the roll
val modDice = dice.map { n -> (n + 3).coerceAtMost(6) } // return a new State<Random, Int> generating only 4-5-6 possible dice values
// Roll example
val d1 = modDice.runA(Random())
```

### CODE REVIEW

**Comment by arturogutierrez** [comment]
PR updated! :)

File: katz/src/main/kotlin/katz/data/State.kt [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+package katz
+
+/**
+ * State[S, A] is basically a function S => (S, A),
+ * where S is the type that represents your state
+ * and A is the result the function produces. In addition
+ * to returning the result of type A, the function
+ * returns a new S value, which is the updated state.
+ */
+class State<S, out A>(val runF: (S) -> Pair<S, A>) {
```
Comment (ffgiraldez) [inline katz/src/main/kotlin/katz/data/State.kt]: please use Tuple2 instead of Pair

Comment (ffgiraldez) [inline katz/src/main/kotlin/katz/data/State.kt]: make it implement `HK2<State.F,S,A>`

Comment (arturogutierrez) [inline katz/src/main/kotlin/katz/data/State.kt]: AFAIK, StateT is who has to support HKT, right? This State is using Identity as Monad Environment, isn't?

Comment (arturogutierrez) [inline katz/src/main/kotlin/katz/data/State.kt]: About Tuple2, I will change it :)

**Comment by raulraja** [comment]
Now that we have `Id` instead of `State` this type should be `StateT`. StateT has three type args. 
`F`, `S` and `A` where F is some other monad. once StateT is implemented you can derive `State` by 
`type State[S, A] = StateT[Id, S, A]`.
StateT allows us to use state in other context beside a plain environment such as `Id` for example `Future`.

File: katz/src/main/kotlin/katz/data/State.kt (hunk 1 above)
Comment (arturogutierrez) [inline katz/src/main/kotlin/katz/data/State.kt]: Raul's comment was very clear 😄 
I will try to implement `StateT` 💪

File: katz/src/main/kotlin/katz/data/State.kt [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    fun run(s: S): Tuple2<S, A> {
+        return runF(s)
+    }
+
+    fun runA(s: S): A = run(s).b
+
+    fun runS(s: S): S = run(s).a
+
+    fun <B> map(f: (A) -> B): State<S, B> {
+        return State { s1 ->
```
Comment (JorgeCastilloPrz) [inline katz/src/main/kotlin/katz/data/State.kt]: you can remove this return by using `fun <B> map(f: (A) -> B) = State<S, B> { ... }` syntax

**Comment by raulraja** [comment]
@arturogutierrez take a look at how `Reader` integrates with `Kleisli` now in master, you are gonna have to do something similar for StateT and State.

**Comment by ffgiraldez** [comment]
I will gonna update this PR soon

**Comment by ffgiraldez** [comment]
i see that map2 and map2Eval was defined here in [cats](https://github.com/typelevel/cats/blob/master/core/src/main/scala/cats/data/StateT.scala#L274)

Now that @raulraja are pushing laws i will not do more test on typeclasses and focus on help him to finish it.

**Pull request merged**